streamdeck.go

```diff
@@ -287,47 +212,155 @@ func (sd *StreamDeck) FillColor(btnIndex, r, g, b int) error {
+		'\x00', '\x00', '\xC4', '\x0E', '\x00', '\x00', '\x00', '\x00',
+		'\x00', '\x00', '\x00', '\x00', '\x00', '\x00'}
+
+	for row := 0; row < c.ButtonSize; row++ {
+		for line := c.ButtonSize - 1; line >= 0; line-- {
+			r, g, b, _ := img.At(line, row).RGBA()
+			imgBuf = append(imgBuf, byte(b), byte(g), byte(r))
+		}
+	}
+	return imgBuf, nil
+}
+
 // FillImage fills the given key with an image. For best performance, provide
 // the image in the size of 72x72 pixels. Otherwise it will be automatically
 // resized.
 func (sd *StreamDeck) FillImage(btnIndex int, img image.Image) error {
-	if err := checkValidKeyIndex(btnIndex); err != nil {
+	if err := sd.checkValidKeyIndex(btnIndex); err != nil {
 		return err
 	}
 
 	// if necessary, rescale the picture
 	rect := img.Bounds()
-	if rect.Dx() != ButtonSize {
-		img = resize(img, ButtonSize, ButtonSize)
+	if rect.Dx() != sd.config.ButtonSize {
+		img = resize(img, sd.config.ButtonSize, sd.config.ButtonSize)
 	}
 
-	imgBuf := make([]byte, 0, ButtonSize*ButtonSize*3)
+	imgBuf, err := sd.encodeImage(img)
+	if err != nil {
+		return err
+	}
 
-	for row := 0; row < ButtonSize; row++ {
-		for line := ButtonSize - 1; line >= 0; line-- {
-			r, g, b, _ := img.At(line, row).RGBA()
-			imgBuf = append(imgBuf, byte(r), byte(b), byte(g))
+	sd.lock.Lock()
+	defer sd.lock.Unlock()
+
+	if sd.config.ImageFormat == "bmp" {
+		err := sd.sendOriginalSingleMsgInLock(btnIndex, 1, imgBuf[0:7776])
+		if err != nil {
+			return err
 		}
+
+		return sd.sendOriginalSingleMsgInLock(btnIndex, 2, imgBuf[7776:])
 	}
 
-	page1 := imgBuf[0 : numFirstMsgPixels*3]
-	page2 := imgBuf[numFirstMsgPixels*3:]
+	headerSize := 8
+	bytesLeft := len(imgBuf)
+	pos := 0
+	pageNumber := uint16(0)
+
+	for bytesLeft > 0 {
+		imgToSend := min(bytesLeft, 1024-headerSize)
+
+		buf := make([]byte, 1024)
+		bytesLeft -= imgToSend
 
-	sd.Lock()
-	sd.writeMsg1(btnIndex, page1)
-	sd.writeMsg2(btnIndex, page2)
-	sd.Unlock()
+		buf[0] = 0x02
+		buf[1] = 0x07
+		buf[2] = byte(sd.config.fixKey(btnIndex))
+		if bytesLeft == 0 {
+			buf[3] = 1
+		} else {
+			buf[3] = 0
+		}
+		binary.LittleEndian.PutUint16(buf[4:], uint16(imgToSend))
+		debug("x %x %x", buf[4], buf[5])
+		binary.LittleEndian.PutUint16(buf[6:], pageNumber)
+
+		copy(buf[8:], imgBuf[pos:(pos+imgToSend)])
+
+		debug("going to Write len(buf): %d imgToSend: %d bytesLeft: %d pageNumber: %d len(imgBuf): %d", len(buf), imgToSend, bytesLeft, pageNumber, len(imgBuf))
+
+		n, err := sd.device.Write(buf)
+		if err != nil {
+			return err
+		}
+		if n != len(buf) {
+			return fmt.Errorf("only wrote %d of %d", n, len(buf))
+		}
+
+		pageNumber++
+		pos += imgToSend
+
+	}
 
 	return nil
 }
 
+func (sd *StreamDeck) sendOriginalSingleMsgInLock(btnIndex int, pageNumber uint16, data []byte) error {
+	buf := make([]byte, 8191)
+	buf[0] = 0x02
+	buf[1] = 0x01
+	binary.LittleEndian.PutUint16(buf[2:], pageNumber)
+	if pageNumber == 2 {
+		buf[4] = 1
+	}
+	buf[5] = byte(sd.config.fixKey(btnIndex))
```

Comment:
I don't think this fixKey is needed anymore. Just using btnIndex selects the correct button

Reply:
with these two changes, the sending images to the original streamdeck works again.
 
<img width="1205" alt="Bildschirmfoto 2025-07-04 um 01 44 58" src="https://github.com/user-attachments/assets/cf3921bd-fa26-4302-aabd-89adf99edc86" />

Reply:
ok. It make sense to have a common numbering schema amongst all streamdecks.